skipper.go

```diff
@@ -1365,11 +1371,49 @@ func run(o Options, sig chan os.Signal, idleConnsCH chan struct{}) error {
 
 	var swarmer ratelimit.Swarmer
 	var redisOptions *skpnet.RedisOptions
+	log.Infof("enable swarm: %v", o.EnableSwarm)
 	if o.EnableSwarm {
 		if len(o.SwarmRedisURLs) > 0 {
 			log.Infof("Redis based swarm with %d shards", len(o.SwarmRedisURLs))
+
+			var cb func(quit chan struct{}, ch chan<- []string)
+			if o.KubernetesRedisNamespace != "" && o.KubernetesRedisName != "" {
+				log.Infof("Use %s/%s to update redis shards", o.KubernetesRedisNamespace, o.KubernetesRedisName)
+				var kdc *kubernetes.Client
+				for _, dc := range dataClients {
+					if kc, ok := dc.(*kubernetes.Client); ok {
+						kdc = kc
+						break
+					}
+				}
+				log.Infof("%s/%s kdc != nil: %v", o.KubernetesRedisNamespace, o.KubernetesRedisName, kdc != nil)
+
+				old := len(o.SwarmRedisURLs)
+				cb = func(quit chan struct{}, w chan<- []string) {
+					ticker := time.NewTicker(10 * time.Second)
+					old := old
+					kdc := kdc
+					for range ticker.C {
+						select {
+						case <-quit:
+							close(w)
+							return
+						default:
+						}
+						a := kdc.GetEndpointAdresses(o.KubernetesRedisNamespace, o.KubernetesRedisName)
+						log.Infof("Redis updater called and found %d", len(a))
+
+						if old != len(a) && len(a) != 0 {
+							log.Infof("Redis updater updating %d != %d", old, len(a))
+							old = len(a)
+							w <- a
+						}
+					}
+				}
+			}
```

Comment:
Lets extract to the helper method

Reply:
refactored to a different implementation